### ISSUE TITLE
Which layer should I pick?

### ISSUE THREAD

**Issue by Karol-G** [kazuto1011/grad-cam-pytorch#17]
Hi,

I integrated Grad-CAM into the vqa-model pythia. My current problem is that I don't know which layer I should pick to generate the heatmap from. Pythia uses a resnet152 and I used `print(*list(model.named_modules()), sep='\n')` to print the layer names of this resnet152. 

<details><summary>**Click to open layer names**</summary>
<p>

> ('7', Sequential(
  (0): Bottleneck(
    (conv1): Conv2d(1024, 512, kernel_size=(1, 1), stride=(1, 1), bias=False)
    (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
    (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (conv3): Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False)
    (bn3): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU(inplace=True)
    (downsample): Sequential(
      (0): Conv2d(1024, 2048, kernel_size=(1, 1), stride=(2, 2), bias=False)
      (1): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
  )
  (1): Bottleneck(
    (conv1): Conv2d(2048, 512, kernel_size=(1, 1), stride=(1, 1), bias=False)
    (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (conv3): Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False)
    (bn3): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU(inplace=True)
  )
  (2): Bottleneck(
    (conv1): Conv2d(2048, 512, kernel_size=(1, 1), stride=(1, 1), bias=False)
    (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (conv3): Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False)
    (bn3): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU(inplace=True)
  )
))
('7.0', Bottleneck(
  (conv1): Conv2d(1024, 512, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
  (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv3): Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (bn3): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu): ReLU(inplace=True)
  (downsample): Sequential(
    (0): Conv2d(1024, 2048, kernel_size=(1, 1), stride=(2, 2), bias=False)
    (1): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  )
))
('7.0.conv1', Conv2d(1024, 512, kernel_size=(1, 1), stride=(1, 1), bias=False))
('7.0.bn1', BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.0.conv2', Conv2d(512, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False))
('7.0.bn2', BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.0.conv3', Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False))
('7.0.bn3', BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.0.relu', ReLU(inplace=True))
('7.0.downsample', Sequential(
  (0): Conv2d(1024, 2048, kernel_size=(1, 1), stride=(2, 2), bias=False)
  (1): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
))
('7.0.downsample.0', Conv2d(1024, 2048, kernel_size=(1, 1), stride=(2, 2), bias=False))
('7.0.downsample.1', BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.1', Bottleneck(
  (conv1): Conv2d(2048, 512, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv3): Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (bn3): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu): ReLU(inplace=True)
))
('7.1.conv1', Conv2d(2048, 512, kernel_size=(1, 1), stride=(1, 1), bias=False))
('7.1.bn1', BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.1.conv2', Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False))
('7.1.bn2', BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.1.conv3', Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False))
('7.1.bn3', BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.1.relu', ReLU(inplace=True))
('7.2', Bottleneck(
  (conv1): Conv2d(2048, 512, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv3): Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (bn3): BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu): ReLU(inplace=True)
))
('7.2.conv1', Conv2d(2048, 512, kernel_size=(1, 1), stride=(1, 1), bias=False))
('7.2.bn1', BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.2.conv2', Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False))
('7.2.bn2', BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.2.conv3', Conv2d(512, 2048, kernel_size=(1, 1), stride=(1, 1), bias=False))
('7.2.bn3', BatchNorm2d(2048, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True))
('7.2.relu', ReLU(inplace=True))

</p>
</details>

There are seven of these blocks in total, but I inserted only the last one.
My guess is that either '7', '7.2', '7.2.conv3' or '7.2.bn3' are the best 'last layer' to pick. ('7.2.relu' throws an error) I tested each of them and I think '7' gives the best results.
But why is that the case? '7' contains all the other layers and isn't really the end layer I think?
Why wouldn't '7.2.conv3' or '7.2.bn3' be the better layers?
And what is even '7'?

I hope you know the answer to this :)

**Comment by kazuto1011** [comment]
The layer "7" result is the same as both "7.2.relu" and "7.2" ones.

* The printed results show your modules recursively. As you may know, the layer "7.2" is the submodule of the layer "7", not the cascaded one. The "7.2.conv3" and "7.2" are also in the same relationship. Calling the layer "7" means to get the very last [featmaps/grads](https://github.com/kazuto1011/grad-cam-pytorch/blob/97afb3b35d/grad_cam.py#L127-L139) after propagating through the contained submodules. 
* The "7.2.relu" error might be caused by the `inplace=True` option, let's take the "7"/"7.2" instead.
* "7.2.conv3" or "7.2.bn3" contains unrelated values that would be cut off in the following "7.2.relu".

Note that still, we cannot specify functions such as `F.relu()` and `F.tanh()`. In that case, a root module and the last submodule may output different results.

**Comment by Karol-G** [comment]
Thank for the answer! I understand it much better now :)
